File: backend/app/routers/employees.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas.employee import EmployeeCreate, EmployeeUpdate, EmployeeOut
from app.schemas.mobile_number import MobileNumberCreate, MobileNumberOut
from app.services import employee_service

router = APIRouter(prefix="/employees", tags=["employees"])


@router.get("", response_model=list[EmployeeOut])
def list_employees(search: str | None = Query(None), lob: str | None = None, include_deleted: bool = False, db: Session = Depends(get_db)):
    return employee_service.list_employees(db, search=search, lob=lob, include_deleted=include_deleted)


@router.post("", response_model=EmployeeOut, status_code=201)
def create_employee(payload: EmployeeCreate, db: Session = Depends(get_db)):
    return employee_service.create_employee(db, payload)


@router.get("/{employee_id}", response_model=EmployeeOut)
def get_employee(employee_id: uuid.UUID, db: Session = Depends(get_db)):
    return employee_service.get_employee(db, employee_id)


@router.put("/{employee_id}", response_model=EmployeeOut)
def update_employee(employee_id: uuid.UUID, payload: EmployeeUpdate, db: Session = Depends(get_db)):
    return employee_service.update_employee(db, employee_id, payload)


@router.delete("/{employee_id}", response_model=EmployeeOut)
def delete_employee(employee_id: uuid.UUID, db: Session = Depends(get_db)):
    return employee_service.soft_delete_employee(db, employee_id)


@router.post("/{employee_id}/mobile-numbers", response_model=MobileNumberOut, status_code=201)
def add_mobile_number(employee_id: uuid.UUID, payload: MobileNumberCreate, db: Session = Depends(get_db)):
    return employee_service.add_mobile_number(db, employee_id, payload)


@router.delete("/{employee_id}/mobile-numbers/{number_id}", response_model=MobileNumberOut)
def remove_mobile_number(employee_id: uuid.UUID, number_id: uuid.UUID, db: Session = Depends(get_db)):
    return employee_service.remove_mobile_number(db, employee_id, number_id)

File: frontend/src/api/employees.ts
import type {
  Employee,
  EmployeeCreateInput,
  EmployeeUpdateInput,
  EmployeeListFilters,
  MobileNumber,
  MobileNumberCreateInput,
} from "../types/employee";

const BASE_URL = import.meta.env.VITE_API_BASE_URL || "http://localhost:8000";

async function request<T>(path: string, options: RequestInit = {}): Promise<T> {
  const res = await fetch(`${BASE_URL}${path}`, {
    headers: { "Content-Type": "application/json" },
    ...options,
  });

  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.detail || `Request failed (${res.status})`);
  }

  if (res.status === 204) return null as T;
  return res.json() as Promise<T>;
}

export function listEmployees(filters: EmployeeListFilters = {}): Promise<Employee[]> {
  const params = new URLSearchParams();
  if (filters.search) params.set("search", filters.search);
  if (filters.lob) params.set("lob", filters.lob);
  const qs = params.toString();
  return request<Employee[]>(`/employees${qs ? `?${qs}` : ""}`);
}

export function createEmployee(payload: EmployeeCreateInput): Promise<Employee> {
  return request<Employee>("/employees", { method: "POST", body: JSON.stringify(payload) });
}

export function updateEmployee(id: string, payload: EmployeeUpdateInput): Promise<Employee> {
  return request<Employee>(`/employees/${id}`, { method: "PUT", body: JSON.stringify(payload) });
}

export function deleteEmployee(id: string): Promise<Employee> {
  return request<Employee>(`/employees/${id}`, { method: "DELETE" });
}

export function addMobileNumber(employeeId: string, payload: MobileNumberCreateInput): Promise<MobileNumber> {
  return request<MobileNumber>(`/employees/${employeeId}/mobile-numbers`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function removeMobileNumber(employeeId: string, numberId: string): Promise<MobileNumber> {
  return request<MobileNumber>(`/employees/${employeeId}/mobile-numbers/${numberId}`, {
    method: "DELETE",
  });
}

export function updateMobileNumberProjectLabel(numberId: string, projectLabel: string | null): Promise<MobileNumber> {
  return request<MobileNumber>(`/employees/mobile-numbers/${numberId}/project-label`, {
    method: "PUT",
    body: JSON.stringify({ project_label: projectLabel }),
  });
}

export async function importEmployeeSheet(file: File): Promise<any> {
  const formData = new FormData();
  formData.append("file", file);
  const res = await fetch(`${BASE_URL}/employees/import`, { method: "POST", body: formData });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.detail || `Import failed (${res.status})`);
  }
  return res.json();
}